[TG-3581] Better Support for Enums in JBMC

mgudemann opened the pull request:
Introduce secific unwind bounds for copy loops in array `clone` for enumerations.

Commit:
Change source location of jump target in {table|lookup}switch

Before we considered the `code_switch_caset` to belong to the target instruction
which lead to uncoverable goals of the form:

   IF condition 1 then GOTO 1
   ...
1: GOTO 2
   ASSERT false // uncoverable block
   ...
2:

Commit:
Move `add_array_types` to java_utils

Commit:
Replace calls to array.clone in values for Enum

For Enum classes this replaces calls to clone via a specialized clone method
call that allows for explicit unwind bounds depending on the number of enum
elements.

Commit:
Split `add_array_type`, specialize for enums

Commit:
Add `java_enum_array_clone_unwind_handler`

This handler unwinds loops in specialized `.clone` functions for `Enum` arrays.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 8ca52901f74c86296a94cf7c041455f51c3e1a36).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78520669
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: c5690b25aa44e71da26a4ed08e4a01390b4fea86).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78520865
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 1fe1e3790b249a197a57078de627ca68f3d0c9d9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78577776
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2197,7 +2212,7 @@ void java_bytecode_convert_methodt::convert_invoke(
   //   accessible if we assume that its access attribute is "more
   //   accessible" than it actually is.
-  irep_idt id = arg0.get(ID_identifier);
+  const irep_idt id = arg0.get(ID_identifier);
```
`const &`?

romainbrenguier replied:
it's not clear for me what `arg0` represents, if you have an intuition about that can you rename the parameter?

peterschrammel commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+  // Also provide a clone method:
+  // ----------------------------
+
+  const irep_idt clone_name = identifier + +".clone:()Ljava/lang/Object;";
```
`+ +`?

peterschrammel commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -5,7 +5,6 @@
 Author: Daniel Kroening, kroening@kroening.com
 
 \*******************************************************************/
-
```
spurious change?

peterschrammel commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
 
+/// Register in the \p symbol_table a new symbol type
+/// java::array[X] where X is byte, float, int, char...
+void add_array_type(
```
maybe add doxygen for parameters

peterschrammel commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+{
+  symbol_typet symbol_type = to_symbol_type(java_array_type(l).subtype());
+
+  const std::string identifier = enum_name
```
you could assign that directly to an `irep_idt`

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2106,6 +2106,21 @@ void java_bytecode_convert_methodt::convert_invoke(
+  const bool is_enum_values_clone_call =
+    enum_clone_symbol.type.get_bool(ID_enumeration) &&
+    (full_method_name ==
+     (calling_class_name + ".values:()[L" + class_type_name + ";"));
```
define is_enum_values_clone_call closer to where it is used

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2220,7 +2235,13 @@ void java_bytecode_convert_methodt::convert_invoke(
+  if(is_enum_values_clone_call)
+  {
+    const std::string clone_name =
+      "java::array[" + calling_class_name + "].clone:()Ljava/lang/Object;";
```
I don't think you can add to a string literal. You should use `std::string("java::array[") + ...` instead

romainbrenguier commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+  code_blockt clone_body;
+
+  code_declt declare_cloned(local_symexpr);
+  clone_body.move_to_operands(declare_cloned);
```
code_blockt has a move method

mgudemann replied:
will change, but this is old code that just moved

romainbrenguier commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+
+  member_exprt new_base_class(
+    new_array, base_class_component.get_name(), base_class_component.type());
+  address_of_exprt retval(new_base_class);
```
const?

romainbrenguier commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+    plus_exprt(old_data, index_symexpr), old_data.type().subtype());
+  dereference_exprt new_cell(
+    plus_exprt(new_data, index_symexpr), new_data.type().subtype());
+  code_assignt copy_cell(new_cell, old_cell);
```
const?

romainbrenguier commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -446,184 +445,222 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+}
+
+/// Register in the \p symbol_table a new symbol for an enum type
+/// \param enum_type_name: Name of the enum type.
```
say that it contains the "java::" prefix?

romainbrenguier commented on jbmc/src/java_bytecode/java_enum_array_clone_unwind_handler.cpp:
```diff
@@ -0,0 +1,65 @@
+
+#include <java_bytecode/java_utils.h>
+
+/// Unwind handler that special-cases the clone functions of arrays of
```
specializes?

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2217,7 +2235,13 @@ void java_bytecode_convert_methodt::convert_invoke(
+  if(is_enum_values_clone_call)
+  {
+    const std::string clone_name =
+      "java::array[" + class_type_name + "].clone:()Ljava/lang/Object;";
```
I don't think you can add to a string literal

mgudemann replied:
I am not sure I understand what you mean, it seems to work without a problem.

smowton committed:
Interpreter: deal with member-of-constant-struct expressions

It could already do index-of-constant-array -- by replacing its custom code with simplify_expr
we can expand that to also support member-of-constant-struct, as occurs when symex's deref'ing
and then the interpreter's environment translate a variable-length array type like int[obj->length]
into int[particular_object.length] and then int[{.length = 1, .data = &xyz}.length].

smowton reviewed (changes requested):
It strikes me this can be done more simply if we can expose a bit of symex's state to the loop unwind handlers (called here https://github.com/diffblue/cbmc/blob/develop/src/cbmc/symex_bmc.cpp#L119, and implemented here https://github.com/diffblue/cbmc/blob/develop/jbmc/src/jbmc/jbmc_parse_options.cpp#L508)

Basically we want to inspect a little bit of the calling context, to check whether we're in `clone()`, and if so whether `this->length` (the length of the array being cloned) is a constant.

Possible approaches:
* Give the unwind handler the ability to call `symext::clean_expr` (which among other things performs constant propagation), so it can directly ask symex for `this->length` and find out whether we know a bound
* Annotate the loop! When generating `.clone()` we know the actual loop length is bounded and that bound is `this->length`. Then symex can evaluate the given bound and give it to the unwind handler.

I think either of these would be much simpler and cleaner than method cloning just so we have a unique function ID key to associate with the decision to unwind.

Commit:
Clarify in convert methods

Change argument names to more meaningful identifiers, lint code

mgudemann commented:
@smowton We decided to go with this solution, look at the outcome and look at the alternative in a new issue.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: d449275c7a8077f606204bccdaf30156d5e10475).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78928253
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

tautschnig reviewed (commented):
I acknowledge that this is work-in-progress, but just to jot down my expectations anyway: 1) there need to be tests added; 2) a change to the interpreter has been added, but it's not clear to me how that relates to the PR subject.

romainbrenguier commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -445,3 +445,224 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+  const auto &index_symexpr = index_symbol.symbol_expr();
+
+  code_declt declare_index(index_symexpr);
+  clone_body.move_to_operands(declare_index);
```
clone_body.add(code_declt(...

mgudemann replied:
done

mgudemann replied:
done

romainbrenguier commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -445,3 +445,224 @@ const std::unordered_set<std::string> cprover_methods_to_ignore
+
+  dereference_exprt old_cell(
+    plus_exprt(old_data, index_symexpr), old_data.type().subtype());
+  dereference_exprt new_cell(
```
const?

mgudemann replied:
done

please note that this code was just moved, not written by me

mgudemann replied:
done, please note that I mainly moved the code and didn't write it

Commit:
Add regression tests for changes to JBMC enumeration support

Update regression tests

mgudemann commented:
@tautschnig please find some regression tests in the latest commit (these were just in TG before)
The interpreter change is to prevent a warning / error message that occurred

tautschnig commented:
> The interpreter change is to prevent a warning / error message that occurred

Would it somehow be possible to explain what this is? The change seems entirely sensible to me, but I still don't really get why it would fit this PR.

mgudemann commented:
@tautschnig this fixes the interpreter when it got a `member-of-constant-struct` like 
`void*[{.length = 6, .data = &dynamic_7_array}.length]`

mgudemann commented:
these occurred with my PR, but I am not sure whether it was the cause or I just paid more attention to the output here

Commit:
Evaluate array index before simplify

Commit:
Mark regression tests as expecting failure for symex driven loading

Currently the regression tests fail with symex driven lazy loading due to a
difference in generated properties.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 9b8237777a037315cb84773756f22efde7024169).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80539334
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

smowton commented:
Proposed alternative implementation here: https://github.com/diffblue/cbmc/pull/2651

mgudemann commented:
superseded by #2651
